packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts

```diff
@@ -0,0 +1,337 @@
+		this.listenTo( this.editor.model.document, 'change:data', () => {
+			const processingStates = this._getProcessingStatesOfDeletedImages();
+
+			processingStates.forEach( processingState => {
+				processingState.controller.abort();
+			} );
+		} );
+	}
+
+	/**
+	 * Gets processing states of images that have been deleted in the mean time.
+	 */
+	private _getProcessingStatesOfDeletedImages(): Array<ProcessingState> {
+		const states: Array<ProcessingState> = [];
+
+		for ( const state of this._processInProgress.values() ) {
+			if ( state.element.root.rootName == '$graveyard' ) {
+				states.push( state );
+			}
+		}
+
+		return states;
+	}
+
+	/**
+	 * Close image editor.
+	 */
+	private _handleImageEditorOnClose() {
+		if ( !this._wrapper ) {
+			return;
+		}
+
+		this._wrapper.remove();
+		this._wrapper = null;
+
+		this.editor.editing.view.focus();
+
+		this.refresh();
+	}
+
+	/**
+	 * Save edited image. In case server respond with "success" replace with edited image,
+	 * otherwise show notification error.
+	 */
+	private _handleImageEditorOnSave( state: ProcessingState, asset: CKBoxRawAssetDefinition ) {
+		const t = this.editor.locale.t;
+		const notification = this.editor.plugins.get( Notification );
+		const pendingActions = this.editor.plugins.get( PendingActions );
+		const action = pendingActions.add( t( 'Processing the edited image.' ) );
+
+		this._processInProgress.set( state.ckboxImageId, state );
+		this._showImageProcessingIndicator( asset );
+
+		this._waitForAssetProcessed( asset.data.id, state.controller.signal )
+			.then(
+				asset => {
+					this._replaceImage( state.element, asset );
+				},
+				() => {
+					if ( !state.controller.signal.aborted ) {
+						notification.showWarning( t( 'Server failed image processing or didn\'t respond' ), {
+							title: t( 'Server failed image processing or didn\'t respond' ),
+							namespace: 'ckbox'
+						} );
+					}
+
+					// Remove processing indicator. It was added only to ViewElement.
+					this.editor.editing.reconvertItem( state.element );
+				}
+			).finally( () => {
+				this._processInProgress.delete( state.ckboxImageId );
+				pendingActions.remove( action );
+			} );
+
+		this.refresh();
+	}
+
+	/**
+	 * Get asset's status on server. If server responds with "success" status then
+	 * image is already proceeded and ready for saving.
+	 */
+	private async _getAssetStatusFromServer( id: string, signal: AbortSignal ): Promise<CKBoxRawAssetDefinition> {
+		const url = new URL( 'assets/' + id, this.editor.config.get( 'ckbox.serviceOrigin' )! );
+		const ckboxEditing = this.editor.plugins.get( CKBoxEditing );
+		const response: CKBoxRawAssetDataDefinition = await sendHttpRequest( {
+			url,
+			signal,
+			authorization: ckboxEditing.getToken().value
+		} );
+		const status = response.metadata!.metadataProcessingStatus;
+
+		if ( !status || status == 'queued' ) {
+			throw new Error( 'Image has not been processed yet.' );
```

Comment:
We should use the `CKEditorError` to be able to distinguish runtime exceptions from known use cases.